### PR TITLE
feat(Core/debug): Add commad to debug lfg

### LINKED ISSUE #5964: Feature: RDF Test/Debug Mode
**Describe your feature request or suggestion in detail**

In the same way that battlegrounds have a `.debug bg` command, a `.debug rdf` command to allow solo testers to enter the RDF queue would be useful. We get a fair number of RDF-related issues reported, and it's impossible to test them without making 5 accounts and running 5 clients, which is a nuisance.

**Describe a possible solution to your Feature or Suggestion in detail**

Current alternatives: there is a solo RDF mod (https://github.com/azerothcore/mod-solo-lfg) but that was described on Discord as 'a little buggy' and may need refactoring. Using a mod to test an issue on a system that might be coded quite differently to that mod doesn't sound very useful in any case, as there would be too much uncertainty about where or why bugs might be caused. A native solution is preferable, but the mod may make writing one much easier.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98576138-feature-rdf-test-debug-mode?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Efymer**
## Changes Proposed:
- As stated here #5964 there's a need to have this command for testers/triage team, it's specially handy for testing events like the Horseman dungeon or the Summer festival dungeon for instance.
- This PR combines the solution from [mod-lfg-solo](https://github.com/azerothcore/mod-solo-lfg) and the solution proposed by @Kitzunu here: https://github.com/azerothcore/azerothcore-wotlk/issues/5964#issuecomment-846156769

Please test it carefully 👐 

## Issues Addressed:

- Closes #5964

## Tests Performed:

- I created a character, enabled the command, tested a few dungeons. 
- Disabled the command and seeing that I was queued in a normal RDF queue.

## How to Test the Changes:

1. Create a character with level 3 of mod permissions and try to queue in a specific/rdf dungeon.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/DungeonFinding/LFGMgr.h [hunk 1]
```diff
@@ -550,6 +550,10 @@ namespace lfg
         static bool HasIgnore(ObjectGuid guid1, ObjectGuid guid2);
         /// Sends queue status to player
         static void SendLfgQueueStatus(ObjectGuid guid, LfgQueueStatusData const& data);
+        // debug lfg command
+        void ToggleTesting();
+        /// For 1 player queue testing
+        bool isTesting() const { return m_Testing; }
```
Comment (FrancescoBorzi) [inline src/server/game/DungeonFinding/LFGMgr.h:556]: please consider `[[nodiscard]]` when applicable

**Comment by temperrr** [comment]
Hi. First time testing PR. I have -NO- experience in this coding language, so I can't check that.

I made a lv 3 gm account char. Tested the command (.debug lfg) . It says <error> in chat.  When joining LFG it does work.
Disabling the command with .debug lfg also gives <error>, but it no longer automatically queues me.

Tested with a normal player in queue aswell. It will show <error>, but will not sent the GM or the Player to the dungeon. When the GM rejoins LFG, he instantly gets the queue. When the player rejoins, the same happens. I think this is intentional.? 

I hope this is what you expected from testing a PR. Please leave feedback on Discord if any more information is required in testing PR's.

**Review by Helias — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
> Hi. First time testing PR. I have -NO- experience in this coding language, so I can't check that.
> 
> I made a lv 3 gm account char. Tested the command (.debug lfg) . When joining LFG it does work.
> Disabling the command with .debug lfg will no longer automatically make the dungeon pop. (correct behaviour)
> 
> Tested with a normal player in queue aswell. but will not sent the GM or the Player to the dungeon. When the GM rejoins LFG, he instantly gets the queue. When the player rejoins, the same happens. I think this is intentional.?
> 
> I hope this is what you expected from testing a PR. Please leave feedback on Discord if any more information is required in testing PR's.

@Efymer

**Comment by pangolp** [comment]
> Hi. First time testing PR. I have -NO- experience in this coding language, so I can't check that.
> 
> I made a lv 3 gm account char. Tested the command (.debug lfg) . When joining LFG it does work.
> Disabling the command with .debug lfg will no longer automatically make the dungeon pop. (correct behaviour)
> 
> Tested with a normal player in queue aswell. but will not sent the GM or the Player to the dungeon. When the GM rejoins LFG, he instantly gets the queue. When the player rejoins, the same happens. I think this is intentional.?
> 
> I hope this is what you expected from testing a PR. Please leave feedback on Discord if any more information is required in testing PR's.

Excellent first report. Well detailed for my taste.

**Comment by LukasVolgger** [comment]
Why no config option?

**Comment by Efymer** [comment]
> @Efymer

@FrancescoBorzi I just talked with @temperrr about the second paragraph.
The same happens when you enable `.debug bg`, instant access for all players regardless if they are gm's. Since this is only ment to be used in development enviroments, I'd say it's completely intentional and it's consistent with the rest of `debug` commands.

Let me know if I need to do something else with this PR :)

**Comment by FrancescoBorzi** [comment]
@Efymer yes, that makes total sense. Just wanted your confirmation as the author of the PR. Good job!

@temperrr thanks a lot for testing!

> Why no config option?
@LukasVolgger this is

File: data/sql/updates/pending_db_world/rev_1624914323095978900.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1624914323095978900');
+
+INSERT INTO `acore_string` (`entry`, `content_default`) VALUES
+  (30096, 'LFG is set to 1 player queue for debugging.'),
+  (30097, 'LFG is set to normal queue.');
+
+INSERT INTO `command` (`name`, `security`, `help`) VALUES
+  ('debug lfg', 3, 'Syntax: .debug lfg\r\nToggle debug mode for lfg. In debug mode GM can start lfg queue with one player.');
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1624914323095978900.sql:8]: @Efymer can you please add `DELETE` before `INSERT` ? So the query will become re-executable

check https://www.azerothcore.org/wiki/sql-standards

**Comment by LukasVolgger** [comment]
> @Efymer yes, that makes total sense. Just wanted your confirmation as the author of the PR. Good job!
> 
> @temperrr thanks a lot for testing!
> 
> > Why no config option?
> 
> @LukasVolgger this is consistent to `.debug bg` , I don't think there is a need to add a config option here. What did you have in mind?

Just for Solo Gaming it would be nice to have because i can enable the option and play solo lfg. So i have to manually debig every time after server restart

File: data/sql/updates/pending_db_world/rev_1624914323095978900.sql (hunk 2 above)
Comment (Efymer) [reply data/sql/updates/pending_db_world/rev_1624914323095978900.sql:10]: Done 👍

**Comment by Efymer** [comment]
> Just for solo gaming it would be nice to have because i can enable the option and play solo lfg. So i have to manually debug every time after server restart

You can also use https://github.com/azerothcore/mod-solo-lfg for that.

**Comment by LukasVolgger** [comment]
@Efymer i know but tbh this could be implemented without that module imo because it's already there for BG...

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Comment by LukasVolgger** [comment]
@Efymer can you please update the PR? Would be a nice feature to have!

**Comment by Efymer** [comment]
Sorry, I completely forgot about the PR. @Kitzunu can you check the changes I just pushed?

**Review by Kitzunu — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
guys, what's the status of this PR? why it says [Ready to be reviewed] if it has already 2 code-review approvals?

please let's not expect imaginary friends to sort PR labels :)

**Comment by IntelligentQuantum** [comment]
Screenshot_20210913_170914

It works fine for me

**Pull request merged**